### PR TITLE
test(invisible): teach the strip property about the blank-filler carve-out

### PR DESCRIPTION
**Pull request by alexander-turner**
Claimed area: `test/invisible.test.mjs` (the `stripInvisible invariants` property + the blank-filler carve-out block).

Unblocks #249, which is unmergeable on this failure and is itself the fix for `main`'s shard-matrix break — so #239, #241 and #244 are queued behind it too.

## The failure

`main` reds intermittently, on whatever PR happens to draw an unlucky seed:

```
not ok 296 - property: stripInvisible invariants
    not ok 6 - STRIP-matchable chars are gone unless preserved by the carve-out
      Property failed after 74 tests
      { seed: -1494323434, path: "73:2:1:9:5:5:6:6:5:7:5:5:6", endOnFailure: true }
      Counterexample: ["ᅟ가"]
```

## It is the test's model that is stale, not the implementation

Characterized on pristine `main` — U+115F/U+1160/U+3164/U+FFA0 are STRIP-matchable and *are* stripped alone or beside Latin, but survive beside a Hangul syllable:

| input | output | filler kept |
| --- | --- | --- |
| `"ᅟ"` | `""` | no |
| `"ᅟa"` | `"a"` | no |
| `"ᅟ가"` | `"ᅟ가"` | **yes** |
| `"가ᅟ"` | `"가ᅟ"` | **yes** |

That is the blank-filler carve-out (`src/invisible.mjs`, "Blank-filler carve-out (Braille / archaic Hangul)") working exactly as designed and already covered by a fixed-case block: a lone filler does real work in archaic Korean, so it is preserved when anchored to a real jamo/syllable, while a *run* — whose only neighbours are fillers — fails the anchor and is stripped. Stripping the anchored one would mangle legitimate content, which is the false positive CLAUDE.md tells this layer to avoid.

The property's allowlist predates that carve-out. It enumerated ZWNJ/ZWJ, kept presentation selectors and a leading BOM, and nothing else — so it contradicted the module's own documented behaviour and its own sibling tests. **No behaviour changes here; `src/` is untouched.**

## Changes

- **The property's allowlist admits an anchored blank filler,** written as the **spec** (`\p{Script=Braille}` non-blank cell; `\p{Script=Hangul}` non-filler) rather than by importing or copying `isPreservedBlankFiller` — a test that calls the implementation to predict the implementation asserts nothing.
- **Two parametrized fixed-case tables, one per script**, each split preserved/stripped. The carve-out was only ever pinned deterministically for U+3164 and only with the filler anchored on **both** sides, so three of the four Hangul fillers and every edge position rode entirely on the seed. Now every filler is pinned in every context.
- **Anchor isolated from position.** Each stripped table holds the string edge fixed and varies only the neighbour's script: `<filler>a` and `a<filler>` must still strip, alongside `a<filler>b` and the bare filler. Without this a bug of the form "preserve any filler at a string edge" would pass the preserved cases unchanged.
- **Braille gets the same table.** `isPreservedBlankFiller` is symmetric across its two branches, so U+2800's edges were exactly as unpinned as Hangul's. It does not diverge from the interior case — worth knowing, so it is pinned rather than assumed.
- **The `Script=Hangul` (not `Script_Extensions`) gate is pinned.** U+3001 IDEOGRAPHIC COMMA and U+30FB KATAKANA MIDDLE DOT are `Script_Extensions=Hangul`; a Japanese sentence containing them must not anchor a filler.
- **The property is seeded with those cases as fixed `examples`.** An anchored filler is a 2-code-point needle in an 80-char haystack, so unseeded fast-check only stumbles onto one occasionally — which is why the counterexample surfaced months later, on an unrelated PR.
- **A negative corpus of ordinary Korean prose** asserting zero findings *and* `countPayloadInvisible === 0`, per CLAUDE.md's rule that every detector carries one. The counter matters separately: one that charged ordinary Hangul would push honest Korean documents over the scattered-invisibles block threshold in `prompt.mjs`.

## Guards proven non-vacuous

| Mutation to `src/invisible.mjs` | Result |
| --- | --- |
| `isPreservedBlankFiller` returns `false` for Hangul fillers (carve-out removed) | `not ok - stripInvisible: blank-filler carve-out` |
| returns `true` unconditionally (fillers preserved with no anchor) | `not ok - property: stripInvisible invariants` — plus 3 other suites |
| **preserve any filler at a string edge** | reds both new tables **and** the property — 9 suites |
| Braille branch `||` → `&&` (edge anchoring lost) | `not ok 20 - the Braille blank, in and out of Braille context` |

The third is the one the reviewer asked for: it is the neighbouring mistake the edge cases exist to rule out, and before this revision the block could not fail for it.

```
FC_REPRODUCIBLE=1 node --test test/invisible.test.mjs   # 315/315
node --test test/invisible.test.mjs  ×4 random seeds    # 315/315 each
pnpm lint                                               # exit 0
```

## Decisions made

- **Fixed the test, not the implementation.** The three candidate readings were (a) strip fillers in Hangul context too, (b) the test's model is wrong, (c) preserve but report. Evidence for (b): the behaviour is documented in the module header, covered by six pre-existing fixed cases, and derives the anchor set from `Script=Hangul` deliberately excluding the fillers. Under (a) real archaic-Korean text loses characters — a false positive in a detection layer, which CLAUDE.md ranks as a real harm.
- **Did not add a report for the preserved filler.** `found` stays empty, consistent with the joiner carve-out: `found` is asserted elsewhere to be exactly the set of categories that *changed the text*, so reporting a preserved character would break that invariant and add a finding operators cannot act on. Under the alternative the sibling property `found non-empty ⇔ the text changed` needs rewriting and every preserved-joiner case emits noise.
- **Spec-restated rather than imported.** Under the alternative (importing `isPreservedBlankFiller`) the property would be a tautology that passes for any implementation.
- **Restructured into per-script tables rather than patching the single edge block** the review suggestion targeted. The suggestion's substance is all here; the wider shape came from finding that the same seed-dependence covered three more fillers. Under the alternative the diff is smaller and U+115F/U+1160/U+FFA0 stay pinned only by luck.

## Lessons Learned

**What:** When a carve-out is added to a detection/transform layer, update the *property* test's allowlist in the same commit as the fixed cases — and pin the fixed cases as a table over every member of the class × every position (start of input, interior, end of input), not one representative in the interior.

**Where:** Any property test in the template that enumerates "what may legitimately survive" — the shape is `const ok = A || B || C; assert.ok(ok, …)`.

**Why:** A property test's allowlist is a second, hand-maintained model of the implementation, and it drifts silently: the carve-out shipped with what looked like thorough fixed-case coverage, and the property still contradicted it. Because the contradiction only surfaces when the generator happens to draw the anchored shape, it presents months later as an intermittent red on an unrelated PR — here it blocked a CI-unbreaking fix and three PRs behind it, and cost a full root-cause investigation to establish that nothing was actually broken.

**What:** A negative marker must vary exactly one variable from the positive case. If the positive is `<filler>가` (edge + anchored) and the negative is `a<filler>b` (interior + unanchored), the pair cannot fail for "preserve anything at an edge" — the nearest wrong implementation.

**Where:** Any preserved-vs-stripped pair in the template's detection tests.

**Why:** Two-variable negative markers read as rigorous and are vacuous against the most likely bug. Holding the position fixed and swapping only the property under test is what makes the assertion load-bearing.

**What:** When a property's interesting input is a short needle in a long generated haystack, pin it via fast-check's `examples` rather than trusting the generator to find it.

**Where:** `fcRunOptions(overrides)` in `test/test-helpers.mjs` already forwards `examples`.

**Why:** The counterexample here was two code points inside an 80-character string. It existed from the day the carve-out landed and took months of unrelated PRs to draw. `examples` converts that into a case that runs on every invocation, seeded or not.

### CODE REVIEW

File: test/invisible.test.mjs [hunk 1]
```diff
@@ -1378,6 +1378,44 @@ describe("stripInvisible: blank-filler carve-out", () => {
...
+        assert.equal(cleaned, input, JSON.stringify(input));
+        assert.deepEqual(found, [], JSON.stringify(input));
+      }
+      // Positive marker: the SAME filler with no Hangul anywhere is stripped, so
+      // the assertions above pin the anchor rule and not a blanket exemption.
+      const orphan = `a${cp(filler)}b`;
+      const { cleaned, found } = stripInvisibleWithReport(orphan);
+      assert.equal(cleaned, "ab", JSON.stringify(orphan));
+      assert.deepEqual(found, [CATEGORY.BLANK_FILLERS], JSON.stringify(orphan));
+    }
```
[the remark this answers is not in the document]
Comment (alexander-turner) [reply test/invisible.test.mjs]: Both correct, and both fixed in `ed81619`. Point 2 especially — the orphan changed the anchor *and* the position, so the block could not fail for the exact neighbouring bug it existed to rule out.

I took the suggestion's substance but restructured rather than patching in place, because the same gap turned out to be wider than the one block: the carve-out was only ever pinned deterministically for U+3164, so the other three fillers' Hangul-context behaviour rode entirely on the seed. Now two parametrized tables, one per script, each split preserved/stripped:

- **Anchor isolated from position.** Every stripped table holds the edge fixed and varies only the neighbour's script: `<filler>a` and `a<filler>` must still strip, alongside `a<filler>b` and the bare filler.
- **Braille gets the same table.** `isPreservedBlankFiller` is symmetric across the two branches, so U+2800 now has preserved-before / preserved-after / preserved-between and the matching stripped edges. It does *not* diverge from the interior case — worth knowing, so it's pinned.
- Plus every filler against a CJK ideograph, ideographic comma and katakana middle dot, which pins the `Script=Hangul` (not `Script_Extensions`) gate — a Japanese sentence must not anchor a filler.
- The property's blank-filler cases are now seeded as fixed `examples` in `fcRunOptions`. An anchored filler is a 2-code-point needle in an 80-char haystack, which is why the original counterexample took months of unrelated PRs to surface.

Non-vacuity, mutating `src/invisible.mjs`:

| Mutant | Result |
| --- | --- |
| preserve any filler at a string edge (the bug you named) | reds both new suites + the property, 9 suites total |
| Braille branch `‖` → `&&` (edge anchoring lost) | `not ok 20 - the Braille blank, in and out of Braille context` |

`FC_REPRODUCIBLE=1` 315/315, plus 4 unseeded runs at 315/315. `src/` still untouched.

---
_Generated by [Claude Code](https://claude.ai/code)_
